Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java

```diff
@@ -166,85 +163,77 @@ public String getJobUuid() {
-                .map(responsePayload ->
-                        new AbstractMap.SimpleImmutableEntry<>(
-                                responsePayload.getState(),
-                                String.join("\n", responsePayload.getLog())))
-                .onErrorResumeNext(err -> {
-                    if (err instanceof SparkJobFinishedException) {
-                        return Observable.error(err);
+        final int GET_JOB_DONE_REPEAT_DELAY_MILLISECONDS = 1000;
+        return getSparkBatchJobRequest()
+                .repeatWhen(ob -> ob.delay(GET_JOB_DONE_REPEAT_DELAY_MILLISECONDS, TimeUnit.MILLISECONDS))
+                .takeUntil(sparkBatchJob -> {
+                    String jobState = getJobState(sparkBatchJob);
+                    return isJobEnded(sparkBatchJob)
+                            || (jobState != null && isDone(jobState));
+                })
+                .filter(sparkBatchJob -> {
+                    String jobState = getJobState(sparkBatchJob);
+                    return isJobEnded(sparkBatchJob)
+                            || (jobState != null && isDone(jobState));
+                })
+                .map(sparkBatchJob -> {
+                    String jobState = getJobState(sparkBatchJob);
+                    if (jobState != null && isDone(jobState)) {
+                        return new AbstractMap.SimpleImmutableEntry<>(
+                                jobState,
+                                getJobLog(sparkBatchJob));
                     } else {
-                        String errHint = "Error getting job status.";
-                        log().warn(errHint + " " + ExceptionUtils.getStackTrace(err));
-                        return Observable.just(new AbstractMap.SimpleImmutableEntry<>("unknown", errHint + err.getMessage()));
+                        return new AbstractMap.SimpleImmutableEntry<>(
+                                sparkBatchJob.schedulerState().toString(),
+                                getJobLog(sparkBatchJob));
                     }
                 });
     }
 
     @Override
     public Observable<String> awaitStarted() {
-        return getResponsePayloadWithState()
-                .retryWhen(error -> error.flatMap(exception -> {
-                    if (exception instanceof SparkJobFinishedException) {
-                        throw propagate(exception);
-                    } else {
-                        // Retry with limited times
-                        return Observable.just(1)
-                                .zipWith(Observable.range(1, getRetriesMax()), (n, i) -> i)
-                                .delay(getDelaySeconds(), TimeUnit.SECONDS);
+        ctrlInfo("The Spark job is starting...");
+        return getSparkBatchJobRequest()
+                .doOnNext(sparkBatchJob -> {
+                    if (sparkBatchJob.properties() != null && sparkBatchJob.properties().sparkMasterUI() != null) {
```

Comment:
The `sparkBatchJob.properties() != null && sparkBatchJob.properties().sparkMasterUI()` can be wrapped as a `getMasterUI(batchResp)` method